src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,297 @@
+            //    4. No partial type declared in the namespace. Otherwise its multiple declaration will
+            //       end up in different namespace.
+            //
+            // - There's no namespace declaration and all types in the document are declared in global namespace:
+            //    1. The cursor is on the name of first declared type.
+            //    2. No partial type declared in the document. Otherwise its multiple declaration will
+            //       end up in different namespace.
+
+            if (namespaceDecls.Length == 1 && compilationUnit.Members.Count == 1)
+            {
+                var namespaceDeclaration = namespaceDecls.Single();
+                Debug.Assert((object)namespaceDeclaration == (object)compilationUnit.Members.Single());
+
+                var shouldTrigger = namespaceDeclaration.Name.Span.IntersectsWith(position) 
+                    && namespaceDeclaration.Name.GetDiagnostics().All(diag => diag.DefaultSeverity != DiagnosticSeverity.Error)
+                    && !(await ContainsPartialTypeWithMultipleDeclarationsAsync(document, namespaceDeclaration, cancellationToken).ConfigureAwait(false));
+
+                return (shouldTrigger, shouldTrigger ? namespaceDeclaration : null);
+            }
+
+            if (namespaceDecls.Length == 0)
+            {
+                var firstMemberDeclaration = compilationUnit.Members.FirstOrDefault();
+
+                var shouldTrigger = firstMemberDeclaration != null
+                    && firstMemberDeclaration.GetNameToken().Span.IntersectsWith(position)
+                    && !(await ContainsPartialTypeWithMultipleDeclarationsAsync(document, compilationUnit, cancellationToken).ConfigureAwait(false));
+
+                return (shouldTrigger, null);
+            }
+
+            return (false, null);
+        }
+
+        /// <summary>
+        /// Try to change the namespace declaration based on the refacoring rules:
+        /// 
+        ///     - if neither declared and target namespace are "" (i.e. global namespace),
+        ///     then we try to change the name of the namespace.
+        ///     - if decalred namespace is "", then we try to move all types declared 
+        ///     in global namespace in the document into a new namespace declaration.
+        ///     - if target namespace is "", then we try to move all members in declared 
+        ///     namespace to global namespace (i.e. remove the namespace declaration).    
+        /// </summary>
+        protected override SyntaxNode ChangeNamespaceDeclaration(
+            SyntaxNode root, 
+            ImmutableArray<string> declaredNamespaceParts, 
+            ImmutableArray<string> targetNamespaceParts)
+        {
+            if (declaredNamespaceParts.IsDefault || targetNamespaceParts.IsDefault || !(root is CompilationUnitSyntax compilationUnit))
+            {
+                return root;
+            }
```

Comment:
it feels weird that our caller woudl even call this if we had a compilation unit.  does VB do the same check here?

same with the part about checking if the parts are IsDfault.

Reply:
i still don't understand this part.  when do you get a non-CompilatoinUnit root node?